### PR TITLE
feat: 支持多模板匹配，任一达到阈值即视为命中

### PR DESCRIPTION
**Pull request by MistEO**
MultiMatcher 暂不支持，仅使用第一个模板

### CODE REVIEW

**Comment by Lancarus** [comment]
文档呢(大声,又不写文档是吧)

**Comment by MistEO** [comment]
> 文档呢(大声),又不写文档是吧

这要啥文档，就是你想象的那样啊

File: src/MaaCore/Config/TaskData.cpp [hunk 1]
```diff
@@ -724,17 +724,48 @@ asst::TaskData::taskptr_t asst::TaskData::generate_match_task_info(std::string_v
...
+    m_templ_required.insert(match_task_info_ptr->templ_names.begin(), match_task_info_ptr->templ_names.end());
 
     // 其余若留空则继承模板任务
-    match_task_info_ptr->templ_threshold = task_json.get("templThreshold", default_ptr->templ_threshold);
+
+    auto threshold_opt = task_json.find("threshold");
+    if (!threshold_opt) {
+        match_task_info_ptr->templ_thresholds = default_ptr->templ_thresholds;
+        match_task_info_ptr->templ_thresholds.resize(match_task_info_ptr->templ_names.size(),
+                                                     default_ptr->templ_thresholds.back());
```
Comment (zzyyyl) [inline src/MaaCore/Config/TaskData.cpp:750]: 怎么直接 .back() 的，不做长度校验然后 copy 吗（

File: src/MaaCore/Config/TaskData.cpp [hunk 2]
```diff
@@ -724,17 +724,48 @@ asst::TaskData::taskptr_t asst::TaskData::generate_match_task_info(std::string_v
...
+    }
+    else if (threshold_opt->is_number()) {
+        match_task_info_ptr->templ_thresholds = { threshold_opt->as_double() };
```
Comment (zzyyyl) [inline src/MaaCore/Config/TaskData.cpp:753]: 这里不扩充吗？不允许多个模板图片一个阈值的写法？

File: src/MaaCore/Vision/MultiMatcher.cpp [hunk 3]
```diff
@@ -14,19 +14,21 @@ using namespace asst;
 
 MultiMatcher::ResultsVecOpt MultiMatcher::analyze() const
 {
-    const auto& [matched, templ, templ_name] = Matcher::preproc_and_match(make_roi(m_image, m_roi), m_params);
+    auto match_results = Matcher::preproc_and_match(make_roi(m_image, m_roi), m_params);
+    const auto& [matched, templ, templ_name] = match_results.front();
```
Comment (status102) [inline src/MaaCore/Vision/MultiMatcher.cpp:18]: 能做成std::vector<std::vector<Result>>么

File: src/MaaCore/Config/TaskData.cpp (hunk 1 above)
Comment (MistEO) [reply src/MaaCore/Config/TaskData.cpp:750]: 前面一行 copy 了啊

File: src/MaaCore/Config/TaskData.cpp (hunk 2 above)
Comment (MistEO) [reply src/MaaCore/Config/TaskData.cpp:753]: 不允许

File: src/MaaCore/Vision/MultiMatcher.cpp (hunk 3 above)
Comment (MistEO) [reply src/MaaCore/Vision/MultiMatcher.cpp:18]: 这是三个东西

Comment (status102) [reply src/MaaCore/Vision/MultiMatcher.cpp:18]: 是指MultiMatcher的结果，做成多模板的每个模板都给一个结果组，而不只是匹配第一个。analyze()或许要改一改

**Comment by Constrat** [comment]
The tasks.json syntax for dark mode / sami mode is this?
```
    "Home": {
        "template": [
            "Terminal.png",
            "TerminalDark.png",
            "TerminalSami.png"
        ],
        "action": "Stop",
        "roi": [
            880,
            100,
            180,
            120
        ]
    },
```

**Comment by Lancarus** [comment]
> The tasks.json syntax for dark mode / sami mode is this? (Like the `"text"`)
> In case of a single template it'll still need the `[` `]` like `"text"`?

Maybe need
```
"templThreshold": [0.8, 0.7, 0.9]
```

**Comment by Constrat** [comment]
> no need, `"template": [""]` & `"template": ""` are both supported.

How should we change the schema? String or Array?
```
"template": {
    "type": "string",
    "default": "",
    "description": "可选项，要匹配的图片文件名\n不填时写默认 \"任务名.png\""
},
```

**Comment by Constrat** [comment]
If the tests go green, good, but I still haven't tested if the templates work correctly.
[30e448b](https://github.com/MaaAssistantArknights/MaaAssistantArknights/pull/6608/commits/30e448b620e1e790a6060e9ac74693134f121307) should have a lot of redundancy because GachaEnter.png, Friends.png, Recruit.png are similar to the dark versions

**Comment by Constrat** [comment]
They went green wow

File: src/MaaCore/Vision/MultiMatcher.cpp (hunk 3 above)
Comment (MistEO) [reply src/MaaCore/Vision/MultiMatcher.cpp:18]: preproc_and_match 把阈值给回来就行了，要改一下，但是我感觉 MultiMatcher 没有多模板的需求，要用再说吧.jpg

Comment (status102) [reply src/MaaCore/Vision/MultiMatcher.cpp:18]: 要不给multi的analyze()加个注释标明一下

**Comment by Constrat** [comment]
@MistEO CN works correctly with the 3 different GUIs (the only problem could be the Store in Sami GUI as it has strange symbols)

There seems to be a big problem for the EN (and probably all the other Overseas)

**Comment by Constrat** [comment]
```
[2023-10-02 08:22:44.507][INF][Px3680][Tx19a8] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":30,"to_be_recognized":["StartUp@StartToWakeUp@LoadingIcon","StartUp@GameStart","StartUp@StartToWakeUp","StartUp@StartToWakeUpOCR","StartUp@StartLoginBServer","StartUp@StartUpConnectingFlag","StartUp","StartUp@CloseAnno","StartUp@CloseAnnoTexas","StartUp@TodaysSupplies","StartUp@NotDownloadVoice","StartUp@OfflineConfirm"]},"first":["StartUpBegin"],"pre_task":"StartUp@StartToWakeUp@LoadingIcon","subtask":"ProcessTask","taskchain":"StartUp","taskid":1}
[2023-10-02 08:22:44.839][INF][Px3680][Tx19a8] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 332 ms , stdout size: 3410660 , socket size: 0
[2023-10-02 08:22:44.873][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@StartToWakeUp@LoadingIcon
[2023-10-02 08:22:44.881][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@GameStart
[2023-10-02 08:22:44.886][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@StartToWakeUp
[2023-10-02 08:22:44.887][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@StartToWakeUpOCR
[2023-10-02 08:22:44.897][TRC][Px3680][Tx19a8] asst::WordOcr [{ : [ 0, 0, 516, 91 ], score: 0.000000 }] by OCR Rec , cost 9 ms
[2023-10-02 08:22:44.897][TRC][Px3680][Tx19a8] Proced []
[2023-10-02 08:22:44.897][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@StartLoginBServer
[2023-10-02 08:22:44.904][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@StartUpConnectingFlag
[2023-10-02 08:22:44.908][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp
[2023-10-02 08:22:44.912][TRC][Px3680][Tx19a8] match_templ | Terminal.png score: 0.997094 rect: [ 907, 129, 123, 60 ] roi: [ 880, 100, 180, 120 ]
[2023-10-02 08:22:44.912][INF][Px3680][Tx19a8] Assistant::append_callback | SubTaskStart {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"MatchTemplate","exec_times":1,"max_times":3,"task":"StartUp"},"first":["StartUpBegin"],"pre_task":"StartUp@StartToWakeUp@LoadingIcon","subtask":"ProcessTask","taskchain":"StartUp","taskid":1,"uuid":"18a1f41ed77967be"}
[2023-10-02 08:22:44.912][TRC][Px3680][Tx19a8] ready to sleep 1000
[2023-10-02 08:22:45.913][TRC][Px3680][Tx19a8] end of sleep 1000
[2023-10-02 08:22:45.913][INF][Px3680][Tx19a8] Assistant::append_callback | SubTaskCompleted {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"MatchTemplate","exec_times":1,"max_times":3,"task":"StartUp"},"first":["StartUpBegin"],"pre_task":"StartUp@StartToWakeUp@LoadingIcon","subtask":"ProcessTask","taskchain":"StartUp","taskid":1,"uuid":"18a1f41ed77967be"}
[2023-10-02 08:22:45.913][TRC][Px3680][Tx19a8] ready to sleep 1000
[2023-10-02 08:22:46.914][TRC][Px3680][Tx19a8] end of sleep 1000

```
Here the "normal" Terminal is detected correctly, but after that during the StartToWakeUp task this happens: 
```
[2023-10-02 08:22:46.914][INF][Px3680][Tx19a8] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":30,"to_be_recognized":["StartUp","StartUp@LoadingText","StartUp@CloseAnno","StartUp@CloseAnnoTexas","StartUp@TodaysSupplies","StartUp@NotDownloadVoice"]},"first":["StartUpBegin"],"pre_task":"StartUp","subtask":"ProcessTask","taskchain":"StartUp","taskid":1}
[2023-10-02 08:22:47.297][INF][Px3680][Tx19a8] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 384 ms , stdout size: 3416936 , socket size: 0
[2023-10-02 08:22:47.327][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp
[2023-10-02 08:22:47.329][ERR][Px3680][Tx19a8] templ size is too large TerminalDark.png image size: 123 60 templ size: 138 80
[2023-10-02 08:22:47.329][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@LoadingText
[2023-10-02 08:22:47.355][TRC][Px3680][Tx19a8] asst::WordOcr [{ Dase: [ 262, 0, 93, 26 ], score: 0.872568 }, { BETA: [ 261, 19, 47, 21 ], score: 0.996800 }] by OCR Pipeline , cost 27 ms
[2023-10-02 08:22:47.355][TRC][Px3680][Tx19a8] Proced []
[2023-10-02 08:22:47.355][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@CloseAnno
[2023-10-02 08:22:47.359][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@CloseAnnoTexas
[2023-10-02 08:22:47.360][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@TodaysSupplies
[2023-10-02 08:22:47.397][TRC][Px3680][Tx19a8] asst::WordOcr [{ TE: [ 189, 11, 53, 42 ], score: 0.449459 }, { 10135: [ 243, 13, 55, 32 ], score: 0.997843 }] by OCR Pipeline , cost 36 ms
[2023-10-02 08:22:47.397][TRC][Px3680][Tx19a8] Proced []
[2023-10-02 08:22:47.397][TRC][Px3680][Tx19a8] asst::PipelineAnalyzer::analyze StartUp@NotDownloadVoice
[2023-10-02 08:22:47.398][TRC][Px3680][Tx19a8] asst::ProcessTask::_run | leave, 43348 ms
[2023-10-02 08:22:47.398][TRC][Px3680][Tx19a8] ready to sleep 1000
[2023-10-02 08:22:48.399][TRC][Px3680][Tx19a8] end of sleep 1000
[2023-10-02 08:22:48.399][TRC][Px3680][Tx19a8] asst::ProcessTask::_run | enter
```
Is it trying to use the Dark Template for the overseas? Why wouldn't it use the "normal" one?

**Comment by Constrat** [comment]
The template size for "Terminal" for all themes in the CN is 138x80 but for the EN it's 123x60.
It takes a screencap with the intention of a smaller template? But then again why is it not using the White default template?

**Comment by Constrat** [comment]
I've changed the EN template and made it bigger (the same dimensions as the CN ones) and now it works both:
1. Doesn't give the [ERR] templ size AND it doesn't use the TerminalDark, but uses the Terminal.png as expected.

CN still works as intended.

**Comment by Constrat** [comment]
To make use of the multi_templ do all the templates need to be the same exact dimensions? Even between different clients? If they do need to be same sizes that'll be a lot more complicated than expected.

**Comment by zzyyyl** [comment]
add line `"cache": false` when use multi-templ will be useful.

**Comment by Constrat** [comment]
> When using multiple templates, adding a line `"cache":false` would be useful.

Ayy that was it. No need to remake all templates.

**Comment by Constrat** [comment]
Oh, unfortunate, the idea of saving performance was nice 😿

**Pull request merged**